flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -4711,6 +4711,11 @@ private void validateSnapshot(SqlNode node, SqlValidatorScope scope, SqlValidato
                         Static.RESOURCE.illegalExpressionForTemporal(
                                 dataType.getSqlTypeName().getName()));
             }
+            if (ns instanceof IdentifierNamespace && ns.resolve() instanceof WithItemNamespace) {
+                // If the snapshot is used over a CTE, then we don't have a concrete underlying
+                // table to operate on. This will be rechecked later in the planner rules.
+                return;
```

Comment:
@tsreaper the fix here just ignore WithItemNamespace. If the ns is not WithItemNamespace, and ns.getTable is null(it may be possible according to your former comment), the validator will still throw NPE exception. 
I think it may be enough to  just ignore when getTable is null(this is  how the validator treat ns.getTable in other method), or it is better to throw a more helpful exception instead of NPE.

Reply:
I understand. If with this fix we still meet a case where `getTable()` is null then this is unexpected and we should take another look. So we should add a `Preconditions.notNull()` here with a message asking users to create a ticket.

Reply:
@wenlong88 and @tsreaper Currently, The table will not be null, because the namespace for snapshot will always be IdentifierNamespace (snapshot must be with a table name, view name or the CTE name).`ns.resolve()` will check `whether resolvedNamespace is null`. But I'm also fine with we add `notNull` check to protect the future change. cc @YuvalItzchakov

Reply:
@godfreyhe What should happen in case it's null? Which exception should be propagated upwards?

Reply:
@YuvalItzchakov Y I think you can throw `CalciteContextException` directly, just call `newValidationError()`

Reply:
@YuvalItzchakov do you want to add null check ?